base/modulus_remainder.h

```diff
@@ -0,0 +1,58 @@
+#ifndef SLINKY_BUILDER_MODULUS_REMAINDER_H
+#define SLINKY_BUILDER_MODULUS_REMAINDER_H
+
+#include <algorithm>
+#include <cassert>
+#include <cstdint>
+
+
+namespace slinky {
+
+// Based on https://github.com/halide/Halide/blob/main/src/ModulusRemainder.h
+/** The result of modulus_remainder analysis. These represent strided
+ * subsets of the integers. A ModulusRemainder object m represents all
+ * integers x such that there exists y such that x == m.modulus * y +
+ * m.remainder. Note that under this definition a set containing a
+ * single integer (a constant) is represented using a modulus of
+ * zero. These sets can be combined with several mathematical
+ * operators in the obvious way. E.g. m1 + m2 contains (at least) all
+ * integers x1 + x2 such that x1 belongs to m1 and x2 belongs to
+ * m2. These combinations are conservative. If some internal math
+ * would overflow, it defaults to all of the integers (modulus == 1,
+ * remainder == 0). */
+struct modulus_remainder {
+    modulus_remainder() = default;
+    modulus_remainder(int64_t m, int64_t r)
+        : modulus(m), remainder(r) {
+    }
+
+    int64_t modulus = 1, remainder = 0;
+
+    // Take a conservatively-large union of two sets. Contains all
+    // elements from both sets, and maybe some more stuff.
+    static modulus_remainder unify(const modulus_remainder &a, const modulus_remainder &b);
```

Comment:
For other set-like types (e.g. `interval_expr`), I've been using `operator|` and `operator&` for union and intersection. Does it make sense to do that here too?

Reply:
Modulus_remainder has many arithmetic ops defined for it, so it might be confusing to use `operator&` for something which is not really arithmetic op? Either way is fine with me, let me know what would you prefer. BTW, `intersection` wasn't used anywhere, so I removed it.

Reply:
Isn't `interval_expr` analogous in all of those ways? That's a set-like thing with arithmetic operators too.